.github/workflows/stackhpc.yml

```diff
@@ -66,13 +66,13 @@ jobs:
           cd $APPLIANCES_ENVIRONMENT_ROOT/terraform
           terraform apply -auto-approve
 
-      - name: Delete infrastructure if failed due to lack of hosts
+      - name: Delete infrastructure if provisioning failed or run cancelled
         run: |
           . venv/bin/activate
           . environments/.stackhpc/activate
           cd $APPLIANCES_ENVIRONMENT_ROOT/terraform
           terraform destroy -auto-approve
-        if: ${{ steps.provision_servers.outcome == 'failure' }}
+        if: always() && steps.provision_servers.outcome == 'failure'
```

Comment:
I don't think this will work if the workflow is cancelled because the second half of the conditional will be false, right? I think removing the `always()` from here and then making the very last step in the workflow (named  'Delete infrastructure' currently) have `if: always()` instead of `success() || cancelled()` is what we actually want?

Reply:
In that case, shouldn't it be `failure()` instead of `always()`? I guess they probably both achieve the same thing in this case, just confused me to see an `always() && ...`

Reply:
> I don't think this will work if the workflow is cancelled because the second half of the conditional will be false, right? I think removing the `always()` from here and then making the very last step in the workflow (named 'Delete infrastructure' currently) have `if: always()` instead of `success() || cancelled()` is what we actually want?

Ignore this, I now understand that you're aiming not to tear down on other failures, so current approach makes sense.

Reply:
Yeah that's it, if ansible fails you can log in via bastion and debug it (recovering secrets and inventory, so you can actually run again locally, log into the monitoring etc).